pydantic/functional_serializers.py

```diff
@@ -282,35 +279,67 @@ def dec(
             when_used=when_used,
             check_fields=check_fields,
         )
-        return _decorators.PydanticDescriptorProxy(f, dec_info)
+        return _decorators.PydanticDescriptorProxy(f, dec_info)  # pyright: ignore[reportArgumentType]
 
-    return dec
+    return dec  # pyright: ignore[reportReturnType]
 
 
-FuncType = TypeVar('FuncType', bound=Callable[..., Any])
+if TYPE_CHECKING:
+    # The first argument in the following callables represent the `self` type:
+
+    ModelPlainSerializer: TypeAlias = Callable[[Any, SerializationInfo], Any]
+    """A model serializer method with the `info` argument, in `plain` mode."""
+
+    ModelPlainSerializerWithoutInfo: TypeAlias = Callable[[Any], Any]
+    """A model serializer method without the `info` argument, in `plain` mode."""
+
+    AnyModelPlainSerializer: TypeAlias = 'ModelPlainSerializer | ModelPlainSerializerWithoutInfo'
+    """A model serializer method in `plain` mode."""
+
+    ModelWrapSerializer: TypeAlias = Callable[[Any, SerializerFunctionWrapHandler, SerializationInfo], Any]
+    """A model serializer method with the `info` argument, in `wrap` mode."""
+
+    ModelWrapSerializerWithoutInfo: TypeAlias = Callable[[Any, SerializerFunctionWrapHandler], Any]
+    """A model serializer method without the `info` argument, in `wrap` mode."""
+
+    AnyModelWrapSerializer: TypeAlias = 'ModelWrapSerializer | ModelWrapSerializerWithoutInfo'
+    """A model serializer method in `wrap` mode."""
+
+    _AnyModelPlainSerializerT = TypeVar('_AnyModelPlainSerializerT', bound=AnyModelPlainSerializer)
+    _AnyModelWrapSerializerT = TypeVar('_AnyModelWrapSerializerT', bound=AnyModelWrapSerializer)
 
 
 @overload
-def model_serializer(__f: FuncType) -> FuncType: ...
+def model_serializer(f: _AnyModelPlainSerializerT, /) -> _AnyModelPlainSerializerT: ...
+
+
+@overload
+def model_serializer(
+    *, mode: Literal['wrap'], when_used: WhenUsed = 'always', return_type: Any = ...
+) -> Callable[[_AnyModelWrapSerializerT], _AnyModelWrapSerializerT]: ...
 
 
 @overload
 def model_serializer(
     *,
-    mode: Literal['plain', 'wrap'] = ...,
-    when_used: Literal['always', 'unless-none', 'json', 'json-unless-none'] = 'always',
+    mode: Literal['plain'] = ...,
+    when_used: WhenUsed = 'always',
     return_type: Any = ...,
```

Comment:
Definitely a nit pick, but can we use the same format as above (one line for args)

Reply:
This was formatted by Ruff (probably because this one is longer: `mode: Literal['plain'] = ...` vs `mode: Literal['wrap']`).

Reply:
Ah ok, ignore me :)